runtime_lib/aircpu/channel.cpp

```diff
@@ -0,0 +1,240 @@
+// Copyright (C) 2023, Advanced Micro Devices, Inc. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "air_channel.h"
+#include "air_tensor.h"
+
+#include <iostream>
+
+#define VERBOSE 0
+
+template <typename T, int R>
+static void _air_channel_put(tensor_t<uint64_t, 0> *channel,
+                             tensor_t<T, R> *src, size_t *_offset,
+                             size_t *_size, size_t *_stride) {
+  size_t offset[4] = {0, 0, 0, 0};
+  size_t size[4] = {1, 1, 1, 1};
+  size_t stride[4] = {1, 1, 1, 1};
+  for (int i = 0; i < R; i++) {
+    offset[i] = _offset[i];
+    size[i] = _size[i];
+    stride[i] = _stride[i];
+  }
+
+  // channel->data is an array of pointers to channel_t objects
+  // if the channel is valid, channel->data[0] should be a valid memory address
+  // otherwise, allocate a new channel
+  if (channel->data[0] == 0) {
+    channel_t<T> *new_channel = (channel_t<T> *)malloc(sizeof(channel_t<T>));
+    new_channel->data =
+        (T *)malloc(sizeof(T) * size[3] * size[2] * size[0] * size[1]);
+    new_channel->_is_full = false;
+    channel->data[0] = (uint64_t)new_channel;
+  }
+
+  channel_t<T> *chan = (channel_t<T> *)channel->data[0];
+
+  // wait until the channel is empty
+  while (chan->_is_full)
+    ;
+
```

Comment:
What are the concurrency assumptions about this code?  This would seem to not be very polite to just spin-wait?

Reply:
Fortunately there is no race condition since it is equivalent to
```c++
const auto f = chan->_is_full;
while (f)
    /* If we loop this is for ever :-) */;
```
after loop-invariant code hoisting.

Reply:
`chan->_is_full` is not exactly loop-invariant. The assumption is that there is another async block that has channel.get which will empty the channel and set `chan->_is_full` to false.

Reply:
Yeah, unfortunately, optimizing compilers will tend to treat this as loop invariant without 'volatile'.  Even then, an architecture can optimize memory accesses at runtime without additional memory barrier instructions.  Even then it seems that you are making assumptions about how async blocks are lowered, in particular that they are implemented as independent threads on separate cores with fairness guarantees, which may necessarily be true (or at least may not be performant)!  Perhaps there is an existing async library that we should be targetting here?  How does mlir/llvm lower its async primitives?

Reply:
That makes sense, I'm taking a closer look at how async primitives are lowered in MLIR

Reply:
Volatile would not solve the problem in a way compatible with the C++ memory model anyway since you want to have some causality between the `_is_full` and the rest of the program.
The C++ way to do it using polling (I do not say this is a good idea, but AMD has to sell a lot of CPU to pay our bills :-) ) would be to use some atomics,
for example
```c++
while (std::atomic_ref { chan->_is_full })
  ;
```

Reply:
Thanks for the suggestion! After some discussion with Steve, I implemented synchronization with condition variables. Please check out my newer commits.

Reply:
Just thinking again at the interesting original code sample:
```c++
  while (chan->_is_full)
    ;
```
Since this is a correct C++, it is a race-free program, so `chan->_is_full` cannot change.
Then, since the abstract C++ machine executing a conformant program has to execute a step eventually (basically a side-effect), a C++ program cannot have an infinite loop.
So the compiler can prove `chan->_is_full` is always false and remove this loop. ;-)